packages/@lwc/template-compiler/src/codegen/static-element-serializer.ts

```diff
@@ -89,6 +91,9 @@ function serializeAttrs(element: Element, codeGen: CodeGen): string {
         })
         .forEach(collector);
 
+    // The additional ${2}/${3} are needed when the class attribute contains static values and for SSR.
+    // In SSR, the style tokens need to be provided for serialization to work, see buildParseFragmentFn.
+    // For non-SSR, the engine will apply style tokens during patching.
     return attrs.join('') + (hasClassAttr ? '${2}' : '${3}');
```

Comment:
To shed some more light on this comment, when the string is created it will always have the `${2}` appended if there was a class attribute present at compile time.

```html
<div class={className}></div>
```
This will compile to something like this:

```js
const $fragment1 = parseFragment`<div${"c0:"} ${2}></div>`;
```
We need the `${2}` in SSR to get the [synthetic shadow style token](https://github.com/salesforce/lwc/blob/1945dcd448c0e1c1c9b504cf281f4a15ce369071/packages/%40lwc/engine-core/src/framework/template.ts#L268).

This is not strictly required when running in a purely browser environment as [`applyStyleScoping`](https://github.com/salesforce/lwc/pull/4071/files#diff-0f0f1360cc9d177fad1e943a03359f9e949ddfd86c8084e092c4811cbccf1484R112) will handle this for us.

Technically, the token will be applied twice in the browser environment, once from the compiled output and again with `applyStyleScoping`.

This shouldn't be fine though as there can only be one attribute and if there is a discrepency between the two tokens it is a bug we need to fix either way.

The reason I kept the `${2}`/`${3}` was to simplify the serialization logic for SSR, otherwise, we would need to find a place to insert the token values during SSR serialization.

Reply:
If the comment needs more light shed on it, then include that extra light in the comment, too!

Reply:
SSR does not support synthetic shadow, so it should never need to render a scope token attribute (only a class, which corresponds to `*.scoped.css`). So is the `${2}` needed?

Reply:
Also, if `applyStyleScoping` is doing duplicated work, I would file an issue to handle that after this gets merged.

Reply:
@nolanlawson I [updated](https://github.com/salesforce/lwc/pull/4071/files#diff-9d36a520650dfa1dd13874ab3870b6f688303b9e5860066178098aadcf45a860R122-R125) the `buildParseFragmentFn` to add the scoped style token.

Now the scoped style token will be appended when the fragment is assembled in both SSR and browser environments and removed the invocations to `applyStyleScoping`.

Reply:
> SSR does not support synthetic shadow, so it should never need to render a scope token attribute (only a class, which corresponds to *.scoped.css). So is the ${2} needed?

It's not needed in SSR, but I don't think it hurts having it there for now since it'll always evaluate to an empty string.

We can refactor in a separate PR to add some logic not to append it specifically for SSR.